eth2-ssv/beacon/prysmgrpc/prysmgrpc.go

```diff
@@ -0,0 +1,260 @@
+package prysmgrpc
+
+import (
+	"context"
+	"github.com/bloxapp/eth2-key-manager/core"
+	"github.com/bloxapp/ssv/beacon"
+	"github.com/ethereum/go-ethereum/event"
+	ptypes "github.com/gogo/protobuf/types"
+	"github.com/herumi/bls-eth-go-binary/bls"
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/status"
+	"time"
+
+	"github.com/bloxapp/ssv/utils/grpcex"
+)
+
+// prysmGRPC implements Beacon interface using Prysm's beacon node via gRPC
+type prysmGRPC struct {
+	conn               *grpc.ClientConn
+	validatorClient    ethpb.BeaconNodeValidatorClient
+	beaconClient       ethpb.BeaconChainClient
+	nodeClient         ethpb.NodeClient
+	network            core.Network
+	graffiti           []byte
+	blockFeed          *event.Feed
+	highestValidSlot   uint64
+	logger             *zap.Logger
+}
+
+// reconnectPeriod is the frequency that we try to restart our
+// streams when the beacon chain is node does not respond.
+var reconnectPeriod = 5 * time.Second
+
+// New is the constructor of prysmGRPC
+func New(ctx context.Context, logger *zap.Logger, network core.Network, graffiti []byte, addr string) (beacon.Beacon, error) {
+	conn, err := grpcex.DialConn(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &prysmGRPC{
+		conn:               conn,
+		validatorClient:    ethpb.NewBeaconNodeValidatorClient(conn),
+		beaconClient:       ethpb.NewBeaconChainClient(conn),
+		nodeClient:         ethpb.NewNodeClient(conn),
+		network:            network,
+		graffiti:           graffiti,
+		blockFeed:          &event.Feed{},
+		logger:             logger,
+	}
+
+	go func() {
+		logger.Info("start receiving blocks")
+		if err := b.receiveBlocks(ctx); err != nil {
+			logger.Error("failed to receive blocks", zap.Error(err))
+		}
+	}()
+
+	return b, nil
+}
+
+// StreamDuties implements Beacon interface
+func (b *prysmGRPC) StreamDuties(ctx context.Context, pubKeys [][]byte) (<-chan *ethpb.DutiesResponse_Duty, error) {
+	streamDuties, err := b.validatorClient.StreamDuties(ctx, &ethpb.DutiesRequest{
+		PublicKeys: pubKeys,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	dutiesChan := make(chan *ethpb.DutiesResponse_Duty)
```

Comment:
Worried about this pattern with unbuffered chans. This function will assume there is some receiver of the channel that will be always ready and therefore not block the routine. It might be better to get rid of that assumption and instead return an event.Subscription? The function can then take a channel as an input argument that can be used for the subscription, which gives the caller more flexibility. This allows this function to be more maintainable.

e.g.:

```go
func (b *prysmGRPC) StreamDuties(ctx context.Context, pubKeys [][]byte, dutiesChan chan *ethpb.DutiesResponse_Duty) (event.Subscription, error) {
```